### ISSUE TITLE
Separate `FixedPointIterator` class

### ISSUE THREAD

**Issue by jwallwork23** [mesh-adaptation/goalie#244]
Reduces the number of methods on `MeshSeq` & co. Avoid overwriting fpe method so that you can (e.g.) still use Hessian-based in `AdjointMeshSeq`.

Related to #238.

**Comment by ddundo** [comment]
When addressing this, I think it would be neat to merge [`goalie.options`](https://github.com/mesh-adaptation/goalie/blob/main/goalie/options.py) with this new class

**Comment by jwallwork23** [comment]
> When addressing this, I think it would be neat to merge [`goalie.options`](https://github.com/mesh-adaptation/goalie/blob/main/goalie/options.py) with this new class

Please could you explain why you're proposing this? To me, it makes sense to have a separate class to hold options.

**Comment by ddundo** [comment]
> > When addressing this, I think it would be neat to merge [`goalie.options`](https://github.com/mesh-adaptation/goalie/blob/main/goalie/options.py) with this new class
> 
> Please could you explain why you're proposing this? To me, it makes sense to have a separate class to hold options.

I'm not insisting of course, but I think it makes sense since 
- all `goalie.options` are only related to fixed point iteration, so it's a bit of a hassle to have to go back and forth between the two files
- there are only a handful of parameters (4 for Hessian- and +3 for goal-oriented)
- I think it's much simpler to do `fpi = FixedPointIterator(Solver, maxiter=10, element_rtol=rtol)` than
```
parameters = goalie.options.AdaptParameters(
    {
        "maxiter": 10,
        "element_rtol": rtol,
    }
)
fpi = FixedPointIterator(Solver, parameters)
```

**Comment by jwallwork23** [comment]
You make a good argument.

**Comment by jwallwork23** [comment]
To be honest, I think the implementation of the options is way too verbose right now anyway so could do with some refactoring.

**Comment by acse-ej321** [comment]
Not sure this helps, but if you think about doing `fpi = FixedPointIterator(Solver,...)` there are additional variables to pass depending on the `Solver` method being used. For the base `forward_solve`,  it might be only `miniter=3, maxiter=10, element_rtol=rtol` but for `solve_adjoint` or `indicate_errors` there are additional convergence criteria options to set. It seems `FixedPointIterator` mostly just runs a loop with different solver, adapt, and convergence criteria options. For separation of concerns, would it work to even move the convergence criteria options to one of the other input classes?

**Comment by ddundo** [comment]
Thanks @acse-ej321, I was thinking about this too. My raw thoughts are that we could have a "quantity of interest" in Hessian-based adaptation too, but we only use it for this qoi convergence criterium. This would require a separate PR. If we do do this, then both would use the same parameters so no separation would be needed here.

**Comment by jwallwork23** [comment]
> Thanks [@acse-ej321](https://github.com/acse-ej321), I was thinking about this too. My raw thoughts are that we could have a "quantity of interest" in Hessian-based adaptation too, but we only use it for this qoi convergence criterium. This would require a separate PR. If we do do this, then both would use the same parameters so no separation would be needed here.

This is an interesting idea. Alternatively, the QoI check could just be a no-op if `get_qoi` isn't implemented or something.

> Edit: actually sorry, there would still be `estimator_rtol` which is only relevant for goal-oriented

For the error estimator, we could definitely set one up for the Hessian-based case. After all, the reason people use the Hessian is because of its relation to interpolation error.

**Comment by jwallwork23** [comment]
Cutting down the number of options classes or removing them entirely is an attractive idea from the point of view of user-friendliness.

**Comment by acse-ej321** [comment]
> > Thanks [@acse-ej321](https://github.com/acse-ej321), I was thinking about this too. My raw thoughts are that we could have a "quantity of interest" in Hessian-based adaptation too, but we only use it for this qoi convergence criterium. This would require a separate PR. If we do do this, then both would use the same parameters so no separation would be needed here.
> 
> This is an interesting idea. Alternatively, the QoI check could just be a no-op if `get_qoi` isn't implemented or something.
> 
> > Edit: actually sorry, there would still be `estimator_rtol` which is only relevant for goal-oriented
> 
> For the error estimator, we could definitely set one up for the Hessian-based case. After all, the reason people use the Hessian is because of its relation to interpolation error.

I have started defining `get_qoi` for almost all the demo cases, even if applying uniform adaptation - treating it as the quantitative property of the solution field(s) I am interested in assessing the quality of the mesh adaptation on, even if it is not used directly as part of the error estimate or general 'monitor function' determination. This might be a misappropriation or another point for maybe including `get_qoi` in most problem definitions, along with `get_inital_conditions`, etc.?

I almost wonder if you could generalise the convergence criteria to something like an ordered list of tuples containing the user-preferred order of evaluation parameters and their specified tolerance levels, i.e. `[(get_qoi, 1.0e-5), (error_estimate, 1.0e-5), (count_elements, 1.0e-2)] ?

**Comment by jwallwork23** [comment]
> I have started defining `get_qoi` for almost all the demo cases, even if applying uniform adaptation - treating it as the quantitative property of the solution field(s) I am interested in assessing the quality of the mesh adaptation on, even if it is not used directly as part of the error estimate or general 'monitor function' determination. This might be a misappropriation or another point for maybe including `get_qoi` in most problem definitions, along with `get_inital_conditions`, etc.?

Yeah I guess it's also useful to define the QoI under non-goal-oriented cases so you can do comparisons.

> I almost wonder if you could generalise the convergence criteria to something like an ordered list of tuples containing the user-preferred order of evaluation parameters and their specified tolerance levels, i.e. `[(get_qoi, 1.0e-5), (error_estimate, 1.0e-5), (count_elements, 1.0e-2)] ?

I guess we could have a method for adding a quantity we want to check the convergence of, which:
* adds the number of elements by default
* potentially adds the others by default too
* user ability to add more quantities that they care about

Perhaps at some point we'll want to implement multi-goal-oriented adaptation, in which case we would want to generalise this, as there'd be multiple QoIs. The difference there (and potentially above) is that we would want to converge w.r.t. *all* of the QoIs, not just one of them.
